pkg/storegateway/postings_codec.go

```diff
@@ -113,35 +115,44 @@ func isDiffVarintSnappyWithMatchersEncodedPostings(input []byte) bool {
 // prepended with any pending matchers to the result.
 // Returned byte slice starts with codecHeaderSnappyWithMatchers header.
 // Length argument is expected number of postings, used for preallocating buffer.
-func diffVarintSnappyWithMatchersEncode(p index.Postings, length int, pendingMatchers []*labels.Matcher) ([]byte, error) {
+func diffVarintSnappyWithMatchersEncode(p index.Postings, length int, requestMatchers indexcache.LabelMatchersKey, pendingMatchers []*labels.Matcher) ([]byte, error) {
 	varintPostings, err := diffVarintEncodeNoHeader(p, length)
 	if err != nil {
 		return nil, err
 	}
 
 	// Estimate sizes
-	estimatedMatchersLen := encodedMatchersLen(pendingMatchers)
+	reqMatchersLen := varint.UvarintSize(uint64(len(requestMatchers))) + len(requestMatchers)
+	estPendingMatchersLen := encodedMatchersLen(pendingMatchers)
 	codecLen := len(codecHeaderSnappyWithMatchers)
 
+	offsetAfterCodec := codecLen
+
 	// Preallocate buffer
-	result := make([]byte, codecLen+estimatedMatchersLen+snappy.MaxEncodedLen(len(varintPostings)))
+	result := make([]byte, codecLen+reqMatchersLen+estPendingMatchersLen+snappy.MaxEncodedLen(len(varintPostings)))
 
 	// Codec
 	copy(result, codecHeaderSnappyWithMatchers)
 
-	// Matchers size + matchers
-	actualMatchersLen, err := encodeMatchers(estimatedMatchersLen, pendingMatchers, result[codecLen:])
+	// Request matchers
+	offsetAfterReqMatchers := offsetAfterCodec + binary.PutUvarint(result[offsetAfterCodec:], uint64(len(requestMatchers)))
```

Comment:
[nit] I find this line a bit hard to read, because it's not immediately clear that it also writes to `result`.